engine/src/test/java/com/redhat/cloud/notifications/processors/email/EmailAggregatorTest.java

```diff
@@ -114,7 +148,57 @@ void shouldTestRecipientsFromSubscription() {
         User user = result.keySet().stream().findFirst().get();
         assertTrue(user.getEmail().equals("user-2"));
         assertEquals(8, ((LinkedHashMap) result.get(user).get("policies")).size());
+        verifyRecipientsResolverInteractions();
+    }
+
+    @Test
+    void shouldTestFallbackOnLegacyRecipientsResolverFetching() {
+
+        featureFlipper.setUseRecipientsResolverClowdappForDailyDigestEnabled(true);
+        // init test environment
+        Application application = resourceHelpers.findApp("rhel", "policies");
+        eventType1 = resourceHelpers.findOrCreateEventType(application.getId(), TestHelpers.eventType);
+        eventType2 = resourceHelpers.findOrCreateEventType(application.getId(), "not-used");
+        resourceHelpers.findOrCreateEventType(application.getId(), "event-type-2");
+        resourceHelpers.createEventTypeEmailSubscription("org-1", "user-2", eventType2, DAILY);
+
+        Endpoint endpoint = new Endpoint();
+        endpoint.setProperties(new SystemSubscriptionProperties());
+        endpoint.setType(EndpointType.EMAIL_SUBSCRIPTION);
+
+        when(endpointRepository.getTargetEmailSubscriptionEndpoints(anyString(), anyString(), anyString(), anyString())).thenReturn(List.of(endpoint));
+
+        when(externalRecipientsResolver.recipientUsers(anyString(), any(), any(), any())).thenThrow(RuntimeException.class);
+
+        when(recipientResolver.recipientUsers(anyString(), any(), any())).then(parameters -> {
+            Set<String> users = parameters.getArgument(2);
+            return users.stream().map(usrStr -> {
+                User usr = new User();
+                usr.setEmail(usrStr);
+                return usr;
+            }).collect(Collectors.toSet());
+        });
+
+        // Test user subscription based on event type
+        Map<User, Map<String, Object>> result = aggregate();
+        verify(emailAggregationRepository, times(1)).getEmailAggregation(any(EmailAggregationKey.class), any(LocalDateTime.class), any(LocalDateTime.class), anyInt(), anyInt());
+        verify(emailAggregationRepository, times(1)).getEmailAggregation(any(EmailAggregationKey.class), any(LocalDateTime.class), any(LocalDateTime.class), eq(0), eq(emailAggregator.maxPageSize));
+        reset(emailAggregationRepository); // just reset mockito counter
+        verify(externalRecipientsResolver, atLeastOnce()).recipientUsers(anyString(), any(), any(), any());
+        verify(recipientResolver, atLeastOnce()).recipientUsers(anyString(), any(), any());
+
+        // nobody subscribed to the right event type yet
+        assertEquals(0, result.size());
+    }
 
+    private void verifyRecipientsResolverInteractions() {
+        if (featureFlipper.isUseRecipientsResolverClowdappForDailyDigestEnabled()) {
+            verify(externalRecipientsResolver, atLeastOnce()).recipientUsers(anyString(), any(), any(), any());
```

Comment:
Can we replace `atLeastOnce()` with `times(1)`?

Reply:
The recipients resolver service will be called on every single event to be aggregated (just in case some forced users are present on the payload).
It can defilitly be improved by adding a simple cache.
I'll add it.